### DISCUSSION TITLE
JobDefs and RunScript resource inheritance

### DISCUSSION THREAD

**Discussion by jschewebbn** [bareos/bareos#2013] [Ideas]
I discovered this while debugging nested JobDefs in the ticket at https://github.com/bareos/bareos/issues/2000
Apparently while settings are passed between nested JobDefs, RunScript resources are not. This seems fundamentally broken to me, however @sebsura  said in the ticket that this is intentional. Can someone explain why this is desired behavior?

I'm a software developer and try and use DRY (don't repeat yourself) in all of my software and IT configurations. So I want to leverage JobDefs to make sure I don't need to duplicate configuration. I've been able to successfully do this for all settings, except RunScript resources due to this.

**Comment by arogge** [comment]
Hi Jon,
we know the current situation with JobDefs is not optimal. We know we should improve the situation, but we don't have an idea what exactly we want to implement and how things should behave.
So right now we don't see anything happening here in the near future.

However, feel free to discuss and maybe even come up with a solution.

**Comment by jschewebbn** [comment]
@arogge  what's the reasoning behind intentionally not inheriting RunScript resources while all other settings are inherited? If I understand your reasons for this I may be able to come up with some reasonable solutions. Pointing to source code and the challenges there would be fine as well.

**Comment by bruno-at-bareos** [comment]
From my point of view, it is really hard to qualify a runscript instruction. Because if your parent jobdefs has a runscript with for example

```
  RunScript {
    Console = "prune stats yes"
    RunsWhen = After
    RunsOnClient = no
  }
```

The consumer may want to add new runscript like those

```
  RunScript {
    Console = "update stats days=3"
    Console = ".bvfs_update %i"
    RunsWhen = After
    RunsOnClient = no
  }
```

And the final consumer being a virtual full job need to have this one 
```
  Run Script {
        console = "update jobid=%i jobtype=A"
        Runs When = After
        Runs On Client = No
        Runs On Failure = No
  }
```

At each stage, you can't easily determine if the new instruction are replacement, or additional ones.

**Comment by jschewebbn** [comment]
Based on https://docs.bareos.org/Configuration/Director.html#config-Dir_Job_JobDefs and the warning there I would assume that Run Script resources are always additive and there is no way to replace them. For me this is the desired behavior and the documentation suggests to me that's how it should be behaving right now. I would argue that the current behavior is in conflict with the documentation.

**Comment by bruno-at-bareos** [comment]
Hello, the documented behavior is actually what's happening.

For example if in a jobdefs I set this runscript part
```
  RunScript {
    Console = ".bvfs_update jobid=%i"
    RunsWhen = After
    RunsOnClient = no
  }
```

And I have the following additional runscript defined in the job
```
  # Use optimized dump script
  RunScript {
    Command = "/usr/bin/pg_dump --format=custom --encoding=SQL_ASCII --clean --if-exists --exclude-table-data='public.pathhierarchy' --exclude-table-data='public.pathvisibility' --file=/var/lib/bareos/bareos.pgdump bareos"
    RunsWhen = "before"
    RunsOnClient = No
  }
  # Update stats run them on server.
  RunScript {
    Console = "prune stats yes"
    Console = "update stats days=3"
    Console = ".bvfs_update"
    RunsWhen = After
    RunsOnClient = no
  }
  # Implement mailx bootstrap after local write
  # The bsr being written to disk before we send a copy by mail
  RunScript {
    RunsWhen = After
    RunsOnClient = no
    Command = "/usr/bin/mailx -s \"Bareos: bootstrap for Job %j\" -a /var/lib/bareos/storage/bsr/%c.%n.bsr bareos@localhost"
  }
```

The final joblog contain all 5 `runscript` instructions
```
 2024-12-18 10:31:15 bareos-dir JobId 18258: shell command: run BeforeJob "/usr/bin/pg_dump --format=custom --encoding=SQL_ASCII --clean --if-exists --exclude-table-data='public.pathhierarchy' --exclude-table-data='public.pathvisibility' --file=/var/lib/bareos/bareos.pgdump bareos"
...
 2024-12-18 10:39:42 bareos-dir JobId 18258: console command: run AfterJob "prune stats yes"
 2024-12-18 10:39:42 bareos-dir JobId 18258: console command: run AfterJob "update stats days=3"
 2024-12-18 10:39:43 bareos-dir JobId 18258: console command: run AfterJob ".bvfs_update"
 2024-12-18 10:39:43 bareos-dir JobId 18258: shell command: run AfterJob "/usr/bin/mailx -s "Bareos: bootstrap for Job catalog.2024-12-18_10.31.13_46" -a /var/lib/bareos/storage/bsr/bareos-fd.catalog.bsr bareos@localhost"
 2024-12-18 10:39:43 bareos-dir JobId 18258: console command: run AfterJob ".bvfs_update jobid=18258"
```

We still agree there's still the opened issue#2000 but the actual documentation seems to reflect the actual behavior of the software.

**Comment by jschewebbn** [comment]
I understand in your comment https://github.com/bareos/bareos/discussions/2013#discussioncomment-11374410 that you are concerned that some may believe that a RunScript element can be replaced by a RunScript in another JobDef. However the documentation clearly states that when taking the settings from a JobDef and applying it to a Job the RunScript elements only add, never replace. Based on that it seems fairly logical that if a JobDef is applied to another JobDef that any RunScript elements would be added and never replace existing ones. I see no documentation to suggest otherwise, thus my question about why the different behavior when a JobDef inherits from another JobDef.

My understanding of issue #2000 is that the order of parsing the files isn't well defined. I've worked around that with the suggestion to put all JobDefs in a single file and order them properly in that file. However that does not appear to have anything to do with why RunScript elements are not added when inheriting from another JobDef. 

If you're going to explicitly disallow RunScript elements from being added by inheriting from another JobDef, I think that should be documented in much the same way you document that RunScript elements are added when a Job inherits from a JobDef.

**Reply by bruno-at-bareos** [reply]
Would you like to propose a PR adapting the documentation in that sense ?

**Comment by jschewebbn** [comment]
Honestly I'd like to propose a PR that changes the code to inherit RunScript elements between JobDefs in the same way that they are inherited from a JobDef to a Job. Once one sees the documentation side by side it just makes it clear to me that they are in conflict with each other.

**Reply by jschewebbn** [reply]
I haven't spent much time in the bareos code, however an initial look at the code suggests that changing https://github.com/bareos/bareos/blob/master/core/src/dird/dird_conf.cc#L2401 to
```
  if (res_type == R_JOB || res_type == R_JOBDEFS) {
```
would take care of my concern.

@sebsura  is this the line where you saw that it looks like RunScript elements are explicitly not propagated between inherited JobDef Resources?

**Reply by sebsura** [reply]
Yes, thats the one.

**Reply by jschewebbn** [reply]
@bruno-at-bareos  is it intentional that this line only contains R_JOB? I suspect this was perhaps an oversight when the code was written and the modification I suggest here could be made and then inheritance of JobDefs would match the behavior seen by Jobs referencing a JobDef.

**Reply by bruno-at-bareos** [reply]
> @bruno-at-bareos is it intentional that this line only contains R_JOB? I suspect this was perhaps an oversight when the code was written and the modification I suggest here could be made and then inheritance of JobDefs would match the behavior seen by Jobs referencing a JobDef.

Not sure why it was done 8 years ago @joergsteffens maybe would better know
https://github.com/bareos/bareos/commit/d5c4ba494647d2132fe5b8369dc293c054104e0b

I remember the introduction of dynamic configuration command `add` look like the changes were also related to this.

Your proposal might certainly be a worth to try. Create a PR with the desired changes. We certainly may want to have a systemtest that check the behavior.

**Reply by jschewebbn** [reply]
PR created at https://github.com/bareos/bareos/pull/2097

I built bareos in a container and ran the tests and all tests passed. It's not surprising that the tests passed I doubt anything was looking at this aspect and the change was minimal. I have not added systemtest to check the behavior. I haven't dug into how such a systemtest would be created.

**Reply by bruno-at-bareos** [reply]
We certainly want to have a systemtest (basically a bash script or a gtest) that will check that expected results are there.
